ArchiSteamFarm/Steam/Integration/SteamChatMessage.cs

```diff
@@ -0,0 +1,187 @@
+		internal const byte NewlineWeight = 61; // This defines how much weight a newline character is adding to the output, limitation enforced by Steam
+		internal const byte ReservedContinuationMessageBytes = ContinuationCharacterBytes * 2; // Up to 2 optional continuation characters
+		internal const byte ReservedEscapeMessageBytes = 5; // 2 characters total, escape one '\' of 1 byte and real one of up to 4 bytes
+
+		internal static string Escape(string message) {
+			if (string.IsNullOrEmpty(message)) {
+				throw new ArgumentNullException(nameof(message));
+			}
+
+			return message.Replace("\\", "\\\\", StringComparison.Ordinal).Replace("[", "\\[", StringComparison.Ordinal);
+		}
+
+		internal static async IAsyncEnumerable<string> GetMessageParts(string message, string? steamMessagePrefix = null) {
+			if (string.IsNullOrEmpty(message)) {
+				throw new ArgumentNullException(nameof(message));
+			}
+
+			int prefixBytes = 0;
+			int prefixLength = 0;
+
+			if (!string.IsNullOrEmpty(steamMessagePrefix)) {
+				// We must escape our message prefix if needed
+				steamMessagePrefix = Escape(steamMessagePrefix!);
+
+				prefixLength = steamMessagePrefix.Length;
+
+				string[] prefixLines = steamMessagePrefix.Split(new[] { "\r\n", "\r", "\n" }, StringSplitOptions.None);
+
+				prefixBytes = prefixLines.Where(prefixLine => prefixLine.Length > 0).Sum(Encoding.UTF8.GetByteCount) + ((prefixLines.Length - 1) * NewlineWeight);
+
+				if (prefixBytes > MaxMessagePrefixBytes) {
+					throw new ArgumentOutOfRangeException(nameof(steamMessagePrefix));
+				}
+			}
+
+			const int maxMessageBytes = MaxMessageBytes - ReservedContinuationMessageBytes;
+
+			// We must escape our message prior to sending it
+			message = Escape(message);
+
+			int messagePartBytes = prefixBytes;
+			StringBuilder messagePart = new(steamMessagePrefix);
+
+			Decoder decoder = Encoding.UTF8.GetDecoder();
+			ArrayPool<char> charPool = ArrayPool<char>.Shared;
+
+			using StringReader stringReader = new(message);
+
+			string? line;
+
+			while ((line = await stringReader.ReadLineAsync().ConfigureAwait(false)) != null) {
+				// Special case for empty newline
+				if (line.Length == 0) {
+					if (messagePart.Length > prefixLength) {
+						messagePartBytes += NewlineWeight;
+						messagePart.AppendLine();
+					}
+
+					// Check if we reached the limit for one message
+					if (messagePartBytes + NewlineWeight + ReservedEscapeMessageBytes > maxMessageBytes) {
+						yield return messagePart.ToString();
+
+						messagePartBytes = prefixBytes;
+						messagePart.Clear();
+						messagePart.Append(steamMessagePrefix);
+					}
+
+					// Move on to the next line
+					continue;
+				}
+
+				byte[] lineBytes = Encoding.UTF8.GetBytes(line);
+
+				for (int lineBytesRead = 0; lineBytesRead < lineBytes.Length;) {
+					if (messagePart.Length > prefixLength) {
+						messagePartBytes += NewlineWeight;
+						messagePart.AppendLine();
+					}
```

Comment:
Consider moving appending line earlier, since it happens anyway:
```suggestion
				if (messagePart.Length > prefixLength) {
					messagePartBytes += NewlineWeight;
					messagePart.AppendLine();
				}

				// Special case for empty newline
				if (line.Length == 0) {
					// Check if we reached the limit for one message
					if (messagePartBytes + NewlineWeight + ReservedEscapeMessageBytes > maxMessageBytes) {
						yield return messagePart.ToString();

						messagePartBytes = prefixBytes;
						messagePart.Clear();
						messagePart.Append(steamMessagePrefix);
					}

					// Move on to the next line
					continue;
				}

				byte[] lineBytes = Encoding.UTF8.GetBytes(line);

				for (int lineBytesRead = 0; lineBytesRead < lineBytes.Length;) {
```

Reply:
No, this had to be done several timer per splitted line in the `for` which you removed below. It's now producing incorrect results for when we need to split a single long line, check unit test.

Reply:
We _could_ pull it out of the special case for newline, but apart from reducing nesting and in turn checking that thing twice in case it is NOT an empty newline this doesn't really accomplish much.

Reply:
Just because we can do some change doesn't mean we should - it must always be evaluated whether the change actually **objectively** improves the current codebase and/or solves some issue without producing new ones, this code has unit tests for a reason - it's not a simple code where you decide to move branches up and down without worrying about change in functionality, it's difficult, and I'm saying this as an author of that function.

It's good to have additional pairs of eyes and suggestions, but sadly not every suggestion is worthy to implement. I also had a lot of ideas that I put aside because of that.